biggus/__init__.py

```diff
@@ -2735,20 +2737,55 @@ def process_chunks(self, chunks):
 
 
 class _Elementwise(ComputedArray):
-    def __init__(self, array1, array2, numpy_op, ma_op):
+    """"
+    Represents an elementwise operation applied to one or two input arrays.
+
+    Elementwise opperations are defined by passing an appropriate function,
+    and optionally an equivalent masked_array function. Examples of elementwise
+    operations incuded addition (two inputs) and cos (one input).
+
+    """
+    def __init__(self, array1, array2, numpy_op, ma_op=None):
         array1 = ensure_array(array1)
-        array2 = ensure_array(array2)
-
-        # Broadcast both arrays to the full broadcast shape.
-        # TypeError will be raised if not broadcastable.
-        array1, array2 = BroadcastArray.broadcast_arrays(array1, array2)
-
-        # Type-promotion - The resultant dtype depends on both the array
-        # dtypes and the operation. Avoid using np.find_common_dtype() here,
-        # as integer division yields a float, whereas standard type coercion
-        # rules with find_common_dtype yields an integer.
-        self._dtype = numpy_op(np.ones(1, dtype=array1.dtype),
-                               np.ones(1, dtype=array2.dtype)).dtype
+
+        expected_n_arrays = None
+        if isinstance(numpy_op, np.ufunc):
+            expected_n_arrays = numpy_op.nin
+            if numpy_op.nout != 1:
+                raise TypeError('Dual output Elementwise ufuncs not yet '
+                                'supported.')
+
+        if array2 is None:
+            # Unary elementwise.
+
+            if expected_n_arrays is not None and expected_n_arrays != 1:
+                # Trigger an exception (TypeError).
+                numpy_op(np.array([1], dtype=array1.dtype), None)
+
+            # Catch warnings for dtype calculation. This typically occurs with
+            # operations which don't work well with 1 (e.g. arctanh).
+            with warnings.catch_warnings(record=True):
+                dtype = numpy_op(np.array([1], dtype=array1.dtype)).dtype
+        else:
+            if expected_n_arrays is not None and expected_n_arrays != 2:
+                # Trigger an exception (TypeError).
+                numpy_op(np.array([1], dtype=array1.dtype))
+
+            # Dual input elementwise
+            array2 = ensure_array(array2)
+
+            # Broadcast both arrays to the full broadcast shape.
+            # TypeError will be raised if not broadcastable.
+            array1, array2 = BroadcastArray.broadcast_arrays(array1, array2)
+
+            # Type-promotion - The resultant dtype depends on both the array
+            # dtypes and the operation. Avoid using np.find_common_dtype(),
+            # here, as integer division yields a float, whereas standard type
+            # coercion rules with find_common_dtype yields an integer.
```

Comment:
Nope, I'm not changing any functionality, just rationalising some of the existing operator overloads.